doc/examples/cookbook.rst

```diff
@@ -30,3 +30,174 @@ Specific claims can be requested using the Authorization request parameter
 
     # client is oic.oic.Client
     client.construct_AuthorizationRequest(request_args=request_args)
+
+Client Assertions
+-----------------
+
+Client assertions can be used for authentication at the IdP token endpoint in
+the OIDC authorization code flow, rather than a client secret. As a challenging
+example, we'll use authenticating with a Microsoft Azure AD IdP, as this
+additionally involves creating a custom client assertion.
+
+Support for client assertions is provided in the ``do_access_token_request``
+method of the ``oic.oic.Client`` class, using keyword arguments to the method:
+
+::
+
+            kwargs = dict(algorithm="RS256", authn_endpoint='token',
+                          authn_method="private_key_jwt")
+
+The ``authn_method`` parameter initiates authentication with a client assertion.
+The ``algorithm`` and ``authn_endpoint`` parameters are needed by the
+``oic.utils.authn.client.PrivateKeyJWT`` class to properly construct the
+assertion. Also required is a signing key in the client's keyjar. A way to
+accomplish this is illustrated by the ``TestPrivateKeyJWT`` class in
+*tests/test_client.py*
+
+::
+
+    _key = rsa_load(os.path.join(BASE_PATH, "data/keys/rsa.key"))
+    kc_rsa = KeyBundle([{"key": _key, "kty": "RSA", "use": "ver"},
+                        {"key": _key, "kty": "RSA", "use": "sig"}])
+    client.keyjar[""] = kc_rsa
+
+The payload of the resulting JWT can be understood by examining the
+``assertion_jwt`` function in ``oic.utils.authn.client``, shown below. Only the
+``alg`` header claim is included when constructing and signing the JWT.
+
+::
+
+    def assertion_jwt(cli, keys, audience, algorithm, lifetime=600):
```

Comment:
This is probably the only thing I slightly disagree with. I do not like a direct copy of the code in the documentation, since it can go out of sync and become confusing. Maybe just leave that as a reference?